In [1]:
%load_ext autoreload

In [2]:
%autoreload 2

# HatEval Spanish Task B

Put a description

## Load data and exploration

First of all, we need to load the dataset in format CSV to a python DataFrame

In [326]:
from lectura import load_data
df = load_data('../../data/train_es_B.tsv')
df

id                                               text  HS  TR  AG
0     20001  Easyjet quiere duplicar el número de mujeres p...   1   0   0
1     20002  El gobierno debe crear un control estricto de ...   1   0   0
2     20003  Yo veo a mujeres destruidas por acoso laboral ...   0   0   0
3     20004  — Yo soy respetuoso con los demás, sólamente l...   0   0   0
4     20007  Antonio Caballero y como ser de mal gusto e ig...   0   0   0
5     20008  @Zurine3 Cuánto hijo de puta sobra en el mundo...   0   0   0
6     20009  Pablo Casado debería estudiar un máster en inm...   0   0   0
7     20010  vos me haces eso te pego una patada en la cara...   0   0   0
8     20012  #Noticiasdelalinea: Detienen en Guatapanal mot...   0   0   0
9     20013          @MaivePerez Llora te lo.mereces por zorra   1   1   1
10    20014  Padres indocumentados realizaron una huelga de...   0   0   0
11    20015  No sé porque hay personas que piensan que todo...   0   0   0
12    20016  A Tomi Alegre JAKSKKSJSJ — Cállate q tdv estoy...   0   0   0
13    20017  @mspn95 @SC_ESPN Callate gallina puta y cntala...   1   1   1
14    20018  Albondigas! Casi al estilo de las fodongas per...   0   0   0
15    20020  Todos: -#NoTodosLosHombres -PPK, presidente de...   0   0   0
16    20021  El juez cita a declarar a la menor que denunci...   0   0   0
17    20022  VAMOOOO LA PUTA MADRE SE LO MERECÍA TANTO ESFU...   0   0   0
18    20023  Pd a ver si cierran de una puta vez Tv3 que no...   0   0   0
19    20024  Yo estaba claro que esta vida de trabajador in...   0   0   0
20    20025  Mira de vd estoy hasta la polla ya, no aguanto...   0   0   0
21    20026  @enkarny282 @irenejunquera No tienes ni puta i...   1   1   1
22    20027  @PepeGonzalezRP @EsmeSese Diappoler* jajaja cá...   0   0   0
23    20028  ¿Mojigatas, moralinas? ¿Neta? ¿Por estar en co...   0   0   0
24    20029  Si cualquier cosa es violación o acoso, se min...   0   0   0
25    20030  Que divertido que consideres que una piba por ...   0   0   0
26    20031  Odio los viejos que se creen que se merecen to...   0   0   0
27    20032           @elultimopeter Tu eres demasiado perra 😂   1   1   1
28    20034  @Manu_Sainz @elchiringuito El papapapartido im...   0   0   0
29    20035  @Setxuki @Daii__24 callate la puta boca gordo ...   0   0   0
...     ...                                                ...  ..  ..  ..
4439  24968  No eres una perra para advertir enseñando los ...   0   0   0
4440  24969  @Fraanpereyra_ No es culpa miiiia uooo, la zor...   1   1   1
4441  24970  señor @JorgeSharp #alcadedevalparaiso tiene cl...   0   0   0
4442  24971  Sobre el video que he publicado, me informan q...   1   0   1
4443  24972  @LaRetuerka @mivecinayyo @MonederoJC Tú eres u...   0   0   0
4444  24974                            Palestino indocumentado   1   1   0
4445  24975  @CaELiKe @ElChamoSalas YA CALLATE PUTA VIEJA D...   1   1   1
4446  24976  @IrantzuVarela Mezclas conceptos, hablar de ac...   0   0   0
4447  24977  @IrantzuVarela Una puta guerra, es una puta gu...   0   0   0
4448  24978  Las mujeres tb son libres de ligarse a los hom...   0   0   0
4449  24979  @5LauraRomo @CircaGh Soy una victima del acoso...   0   0   0
4450  24980  #OJO Detienen a 95 #indocumentados en desierto...   0   0   0
4451  24981  Lo nuestro nunca iba a funcionar porque tu ere...   0   0   0
4452  24982  @Republicana2018 Callate troll y la concha de ...   0   0   0
4453  24984  @joancach @DaleFiera @mauroboselli Toñito otra...   0   0   0
4454  24985  Que pesadilla mi padre contandoles mi vida y e...   0   0   0
4455  24987  SABES A QUIÉN SE A TIRADO LA PELIRROJA ESTA LA...   1   1   1
4456  24988  @nerayu3107 Cállate la puta boca que me muero ...   0   0   0
4457  24989      @steeficorrea Outfit***** puta cállate jajaja   1   1   1
4458  24990  #LoMásLeído | El Gobierno refuerza con 323 pla...   0   0   0
4459  24991  Tengo una duda: Las mujeres arabes: ¿que deben...   0   0   0
4460  24992  Nao: el robot que ens

Let's see the datset size

In [4]:
df.shape

(4469, 5)

A total of 4469 tweets. We can examinate the distribution of the dataset for binary hate speech

In [5]:
df.groupby('HS').size() / df.shape[0]

HS
0    0.588722
1    0.411278
dtype: float64

* 2631 (59%) no hate speech
* 1838 (41%) hate speech

In [6]:
df[df['HS'] == 1].groupby('AG').size() / 1838

AG
0    0.192057
1    0.807943
dtype: float64

* 353 (19%) no agressive
* 1485 (81%) agressive

In [7]:
df[df['HS'] == 1].groupby('TR').size() / 1838

TR
0    0.392274
1    0.607726
dtype: float64

* 721 (39%) no agressive
* 1117 (61%) agressive

We see if there is any null data

In [8]:
df.isnull().sum()

id      0
text    0
HS      0
TR      0
AG      0
dtype: int64

There isn't any null data

We start seeing general information about the data

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4469 entries, 0 to 4468
Data columns (total 5 columns):
id      4469 non-null int64
text    4469 non-null object
HS      4469 non-null int64
TR      4469 non-null int64
AG      4469 non-null int64
dtypes: int64(4), object(1)
memory usage: 174.6+ KB


In [10]:
# We can list non numerical properties, with a boolean indexing of the Series df.dtypes
df.dtypes[df.dtypes == object]

text    object
dtype: object

#### gsitk preprocessing

In [53]:
from gsitk.preprocess import pprocess_twitter, normalize
import string
from sklearn.base import BaseEstimator, TransformerMixin
from nltk.corpus import stopwords
import re
import emoji
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer
class gsitkTransformer(BaseEstimator, TransformerMixin):
    def __init__(self, stemming):
        self.stemming = stemming
        self.emojis_list = emoji.UNICODE_EMOJI
        self.emoji_pattern = emoji.get_emoji_regexp()
        self.trash = ['“', '”', '—', '’', '–', '«', '»', '・・・', 'tq', '€', '…', '‘', 'elonghashtag', 'âž\x9d', 'ðÿ',
                     '¦', '‹ðÿ\x8f', '€', '„', '¯', 'ðÿ‡\xadðÿ‡·', 'âšï¸\x8f', '‰', 'ðÿ¤', 'ª', '\x8f', 
                     'ðÿ¥\x81ðÿ¥\x81', 'ðÿ˜', '♪', '►', '͡°', '•', 'âž\x9d', '¤', '〝']
    stop = stopwords.words('spanish')
    def tokenize_doc(self, doc):
        text = pprocess_twitter.preprocess(doc).replace('<allcaps>', '').replace('<elong>', '')
        text = text.replace('<hastag>', '<hashtag>').replace('<user>', '<usuario>').replace('<repeat>', '<repetición>')
        text = text.replace('<number>', '<número>').replace('\u2060', '').replace('\u2066', '').replace('¤', '')
        text = text.replace('<', '').replace('>', '').replace('.user', 'user').replace('¿', '').replace('—', '').replace('¡', '')
        text = text.replace('xdd', 'xd').replace('\u2069', '').replace('\u2069', '').replace('…', '').replace('˜', '')
        text = text.replace('ª', '')
        trash = ['“', '”', '—', '’', '–', '«', '»', '・・・', 'tq', '€', '…']        
        tokens = normalize.preprocess(text)
        if self.stemming:
            snow = SnowballStemmer('spanish')
            tokens = [snow.stem(t) for t in tokens]
        tokens = [w for w in tokens if w not in string.punctuation and w not in trash 
                  and w not in self.stop and w not in self.emojis_list
                  and not self.emoji_pattern.search(w)]
        return tokens
    
    def fit(self, docs, y=None):
        return self
    
    def transform(self, docs):
        return [self.tokenize_doc(doc) for doc in docs]        

## Features

### Word Embeddings

In [12]:
from gsitk.features.word2vec import Word2VecFeatures
w2v_extractor = Word2VecFeatures(w2v_model_path='../../../../NOBACKUP/SBW-vectors-300-min5.txt', w2v_format='google_txt', convolution=[1,0,0])

### Simon Features

In [13]:
from lectura import lexicon_generation
lexicon = lexicon_generation(df)

In [498]:
from collections import Counter
# lexicon mejorado
def generate_lexicon(df):
    textrf = gsitkTransformer(False)

    # HS
    hs = df['text'][df['HS'] == 1].values
    hs_tokens = textrf.transform(hs)
    hs_tokens = [word for doc in hs_tokens for word in doc]
    hs_lexicon = []

    contador = Counter(hs_tokens)
    comunes = contador.most_common()

    for tupla in comunes:
        hs_lexicon.append(tupla[0])

    # NO HS
    no_hs = df['text'][df['HS'] == 0].values
    no_hs_tokens = textrf.transform(no_hs)
    no_hs_tokens = [word for doc in no_hs_tokens for word in doc]
    no_hs_lexicon = []

    contador = Counter(no_hs_tokens)
    comunes = contador.most_common()

    for tupla in comunes:
        no_hs_lexicon.append(tupla[0])
    no_hs_lexicon = list(set(no_hs_lexicon) - set(hs_lexicon))
    return [hs_lexicon, no_hs_lexicon]

In [624]:
from gsitk.features import simon
from gensim.models.keyedvectors import KeyedVectors

#embedding_model = KeyedVectors.load_word2vec_format('../../../../NOBACKUP/SBW-vectors-300-min5.txt', binary=False)
simon_model = simon.Simon(lexicon=dev_lexicon, embedding=embedding_model, n_lexicon_words=100)
similarity_model = simon.simon_pipeline(simon_transformer=simon_model, percentile=100)

#### Sentiment lexicon

In [16]:
with open('../../data/positive_words_es.txt') as f:
    positive_words = f.readlines()
positive_words = [word.strip() for word in positive_words]
with open('../../data/negative_words_es.txt') as f:
    negative_words = f.readlines()
negative_words = [word.strip() for word in negative_words]

In [17]:
sentiment_lexicon = [positive_words, negative_words]
sentiment_model = simon.Simon(lexicon=sentiment_lexicon, embedding=embedding_model, n_lexicon_words=250)
senti_model = simon.simon_pipeline(simon_transformer=sentiment_model, percentile=25)

In [18]:
sentiment_vocabulary = [word for sublexicon in sentiment_lexicon for word in sublexicon]

#### Hashtag lexicon

In [19]:
from lectura import hashtag_lexicon_generation
hashtag_lexicon = hashtag_lexicon_generation(df)

### Surface Features

We include the TF-IDF vectorizer

In [20]:
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer, TfidfTransformer
from sklearn.feature_extraction import DictVectorizer
from lectura import spanish_tokenizer

#count_vect = CountVectorizer(ngram_range=(1, 2), stop_words='english')
#X_counts = count_vect.fit_transform(transcription_train)
#X_test = count_vect.transform(transcription_test)
#X_counts.shape
#tfidf_transformer = TfidfTransformer()
#vectors_train = tfidf_transformer.fit_transform(X_counts)
#vectors_test = tfidf_transformer.transform(X_test)
#vectors_train.toarray()

ngrams_featurizer = Pipeline([
  ('count_vectorizer',  CountVectorizer(ngram_range = (2, 3), encoding = 'ISO-8859-1', 
                                        tokenizer=spanish_tokenizer)),
  ('tfidf_transformer', TfidfTransformer())
])

## Classification

The estimators to be used will be:

In [21]:
from sklearn import svm, linear_model, tree, ensemble, naive_bayes, neighbors, gaussian_process, multiclass
import numpy as np
less_than_1 = [.0001,.001,.01,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0]
classifiers = {'logistic_regression': 
                   {'model': linear_model.LogisticRegression(), 
                    'params': 
                    {'clf__penalty': ['l1','l2'], 'clf__tol': [0.0001,0.001,0.01,0.1], 
                     'clf__C': np.append(less_than_1, np.arange(1,16))}}, 
               'linear_SVC': 
                   {'model': svm.LinearSVC(dual=False), 
                    'params': 
                    {'clf__penalty': ['l1','l2'], 'clf__tol': [0.0001,0.001,0.01,0.1], 
                     'clf__C': np.append(less_than_1, np.arange(1,16))}}, 
               'random_forest': 
                   {'model': ensemble.RandomForestClassifier(), 
                    'params': 
                    {'clf__n_estimators': np.arange(10, 110, 10), 'clf__max_depth': np.append(np.arange(1, 11), None)}} 
}

In [22]:
# Training and test spliting
from sklearn import preprocessing

# Evaluation
from sklearn import metrics
from sklearn.model_selection import cross_val_score, KFold, StratifiedKFold, train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score

In [24]:
df.groupby('HS').size()/df.shape[0]

HS
0    0.588722
1    0.411278
dtype: float64

Load test data and exploration in the same way than train data.

In [25]:
df_test = load_data('../../data/dev_es_A.tsv')
df_test

id                                               text  HS  TR  AG
0    20005  Me estoy comiendo la picada árabe más rica de ...   0   0   0
1    20006      @Haryachyzaychyk Callate zorra y mama duro! 😍   1   1   1
2    20011  Acabo de escuchar a Casado diciendo que hay DE...   0   0   0
3    20019  Y NADIE SE HA PREGUNTADO LO QUE LE VA A COSTAR...   1   0   1
4    20033  @Fed_Durand Callate come sobra, más zorra son ...   1   1   1
5    20039   te quiero hacer mi reina árabe bomboncitocaramel   0   0   0
6    20046            @andreaacata cállate perra JAJAJAJAJAJA   1   1   1
7    20047  En "La Reina del Sur" de @perezreverte , el pe...   0   0   0
8    20063  @AnderssonBoscan Vieja zorra imbécil y MENTIRO...   1   1   1
9    20101  @rjimenez_perez tu eres un hijo de puta perver...   0   0   0
10   20111  @CanalusHeresy @Fourty35 @IrantzuVarela A, que...   1   1   1
11   20124  @PoleoRafael Mira hijo de puta mejor cállate l...   0   0   0
12   20127  Una tía no es una guarra por hacer una foto de...   0   0   0
13   20138  Ea pues ya solo falta que Albert Rivera vaya a...   1   1   0
14   20142  @KobaML Claro, es un hoyo(coño) que se abre co...   0   0   0
15   20148  Racismo es que hagas enfrijoladas con frijoles...   0   0   0
16   20161  Cada puta semana igual. La puta carpeta. Estoy...   0   0   0
17   20162  *chica súper sufriendo* 'Cielo, la menstruació...   1   1   1
18   20175  Ay Dios cuando se entere Casado de que la ola ...   1   0   0
19   20188                  @Relatofeminista Hasta el coño xD   0   0   0
20   20210  @adrispears @7santy Tú eres la puta Beyoncé si...   0   0   0
21   20213  @britneyspears Perra te lo mereces, esa presen...   1   1   1
22   20232  @VidalSieteDiez Tú eres la perra, no te dejes ...   1   1   1
23   20241   @PabloAlgeo Cállate la concha de tu hermana puta   1   1   1
24   20255    la puta madre que te re mil parió Rocco callate   0   0   0
25   20265  Tema inmigración de Venezuela como estamos y q...   1   0   0
26   20288                    Tú eres la zorra Antoniaaaaaaaa   1   1   1
27   20302  @palbarjh @edusolanosolano Es correcto una cos...   0   0   0
28   20303  @Ricardo202 Imbecil eres tú. Ya sácate ese ran...   0   0   0
29   20305  @CamposYuliyuli @Pauliitiiz Callate vos sos ca...   1   1   1
..     ...                                                ...  ..  ..  ..
470  24649  yo me pregunto como es el teclado de los que e...   0   0   0
471  24651  @elrod_g @jonawoodz35 Jajajajajajajajajajajaaj...   0   0   0
472  24677  ni laurita ni fede ni florencia jEY MAMMON SE ...   0   0   0
473  24686  No me gusta el titular. Culpabilizar a la inmi...   0   0   0
474  24691  @thebossniki Tu eres un imbécil maleducado que...   0   0   0
475  24692  omegle niña árabe jugando para mí https://t.co...   0   0   0
476  24696  @viviana_bonilla @veronicaariasf Como tú eres ...   1   1   1
477  24701             @FlorenciaJerez8 Cállate perra es mioo   1   1   1
478  24705  @SebasGames_ @Rubixsitx que fan que sos man la...   0   0   0
479  24733  Racismo es no permitir que te eche mis güeros ...   0   0   0
480  24736  Siempre hay una amiga del novio que por más in...   1   1   1
481  24746  @rU_1970 muestra como teni esa vagina de puta ...   1   1   1
482  24768  1.Sera casualidad que cuando hay una gran entr...   1   0   1
483  24773  No hay nada que merezca más la pena, que mi pe...   0   0   0
484  24817  Ya empezamos con las tonterías!!! Lo ha hecho ...   0   0   0
485  24821  Orita mi perra me despertó para enseñarme su n...   0   0   0
486  24822  @Relatofeminista 'A mí sólo me gustan las tías...   1   0   0
487  24845  Otra verguenza que tenemos que soportar los es...   1   0   1
488  24864  rosalia me podría decir sudaca i wouldn't give...   1   1   1
489  24870  @zepedajantonio Tú eres el hijo de perra, voy ...   0   0   0
490  24895  @xSuzuHopkins @gimme5yens El de mi polla a gra...   1   1   1
491  24907  La Policía tramita en Algeciras 400 expediente...   1   0   1
492  24909  

In [26]:
df_test.shape

(500, 5)

In [33]:
df_test.groupby('HS').size() / df_test.shape[0]

HS
0    0.556
1    0.444
dtype: float64

In [34]:
df_test[df_test['HS'] == 1].groupby('AG').size() / 222

AG
0    0.207207
1    0.792793
dtype: float64

In [35]:
df_test[df_test['HS'] == 1].groupby('TR').size() / 222

TR
0    0.382883
1    0.617117
dtype: float64

In [308]:
#vocab_set = set(embedding_model.vocab.keys())
#print(df.shape)
#df = df[df['text'].str.split(' ').apply(lambda toks: len(set(toks) & vocab_set)) > 0]
#print(df.shape)
# Execute this if you use simon or word embeddings (Tweet with problems)
df.drop([2664], inplace=True)

In [330]:
df['text'].values[2665]

'@Manaka_msyra @BrianD_Campbell callate zorra malnacida esta el armario empotrado tmb y tengo mil cosas ahi dentro'

In [338]:
df_train = df[df['HS'] == 1]
df_test = df_test[df_test['HS'] == 1]

In [471]:
# Train Test splitting
X_train = df_train['text'].values
y_train = df_train['TR'].values
X_test = df_test['text'].values
y_test = df_test['TR'].values

In [472]:
print(sorted(Counter(y_train).items()))

[(0, 721), (1, 1117)]


In [473]:
items = Counter(y_train)
diff_train = items[1] - items[0]
diff_train

396

In [474]:
print(sorted(Counter(y_test).items()))

[(0, 85), (1, 137)]


In [475]:
items = Counter(y_test)
diff_test = items[1] - items[0]
diff_test

52

In [544]:
df_evaluation = load_data('../../data/test_es.tsv')
df_evaluation.shape

(1600, 2)

In [545]:
X_evaluation = df_evaluation['text'].values

In [493]:
df_competition_train = load_data('../../data/train_es_B.tsv')
df_competition_test = load_data('../../data/dev_es_B.tsv')
df_competition_train.drop([3356], inplace=True)

In [494]:
import pandas as pd
df_dev = pd.concat([df_competition_train, df_competition_test])
df_dev.shape

(4968, 5)

In [622]:
dev_lexicon = generate_lexicon(df_dev)
dev_hashtags = hashtag_lexicon_generation(df_dev)

In [500]:
df_HS = df_dev[df_dev['HS'] == 1]

#### Oversampling

In [476]:
df_extra = df[df['HS'] == 0]
X_extra = df_extra['text'].values
y_extra = df_extra['TR'].values

In [477]:
X_extra = np.random.choice(X_extra, size=diff_train)
y_extra = y_extra[:diff_train]

In [478]:
X_train = np.append(X_train, X_extra)
y_train = np.append(y_train, y_extra)

In [479]:
assert len(X_train) == len(y_train)
permutation = np.random.permutation(len(X_train))
X_train = X_train[permutation]
y_train = y_train[permutation]

In [480]:
print(sorted(Counter(y_train).items()))

[(0, 1117), (1, 1117)]


In [278]:
df_extra_test = df_test[df['HS'] == 0]
X_extra_test = df_extra_test['text'].values
y_extra_test = df_extra_test['AG'].values
X_extra_test = np.random.choice(X_extra, size=diff_test)
y_extra_test = y_extra[:diff_test]
X_test = np.append(X_test, X_extra_test)
y_test = np.append(y_test, y_extra_test)
assert len(X_test) == len(y_test)
permutation = np.random.permutation(len(X_test))
X_test = X_test[permutation]
y_test = y_test[permutation]

In [481]:
print(sorted(Counter(y_test).items()))

[(0, 85), (1, 137)]


In [634]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.feature_extraction import DictVectorizer
from sklearn.feature_selection import SelectKBest, chi2
from sklearn import metrics
from sklearn.metrics import make_scorer, accuracy_score
from lectura import TextTransformer, LexicalStats, spanish_tokenizer, LowerTransformer, SubjectivityStats
from sklearn.decomposition import LatentDirichletAllocation
from lectura import TwitterStats, DenseTransformer
import emoji
from nltk.tokenize import TweetTokenizer
from sklearn.feature_selection import SelectKBest, SelectPercentile, f_classif, chi2
from sklearn.decomposition import PCA
from sklearn.preprocessing import MaxAbsScaler

tweet_tokenizer = TweetTokenizer(reduce_len=True, strip_handles=True)
gsitk_tokenizer = gsitkTransformer(True)

textFeatures = Pipeline([
    ('features', FeatureUnion([
        ('lexical_stats', Pipeline([
            ('stats', LexicalStats()),
            ('vectors', DictVectorizer()),
            ('scaler', MaxAbsScaler())
        ])),
        ('word_embeddings', Pipeline([
            ('text_to_tokens', gsitkTransformer(False)),
            ('embedding_features', w2v_extractor)
        ])),
        ('simon_features', Pipeline([
            ('text_to_tokens', gsitkTransformer(False)),
            ('similarity', similarity_model)
        ])),
        #('sentiments', Pipeline([
        #   ('text_to_tokens', gsitkTransformer()),
        #   ('sentiment', senti_model)
        #])),
        ('sentiments', TfidfVectorizer(vocabulary=sentiment_vocabulary, tokenizer=tweet_tokenizer.tokenize)),
        ('hashtags', Pipeline([
            ('lowercase_transformation', LowerTransformer()),
            ('hashtag_vectorizer', CountVectorizer(vocabulary=dev_hashtags, tokenizer=tweet_tokenizer.tokenize))            
        ])),
        ('twitter stats', Pipeline([
            ('stats', TwitterStats()),
            ('vectors', DictVectorizer())
        ])),
        ('emojis', TfidfVectorizer(token_pattern=emoji.get_emoji_regexp(), tokenizer=tweet_tokenizer.tokenize)),
        ('words', TfidfVectorizer(tokenizer=gsitk_tokenizer.tokenize_doc)),
        ('ngrams', ngrams_featurizer),
        ('character_ngram', TfidfVectorizer(analyzer='char', ngram_range=(2,7))),
        #('subjectivity_stats', Pipeline([
        #    ('stats', SubjectivityStats()),
        #    ('vectors', DictVectorizer())
        #])),
        ('lda', Pipeline([
            ('count', CountVectorizer(tokenizer=gsitk_tokenizer.tokenize_doc)),
            ('lda', LatentDirichletAllocation(n_topics=10, max_iter=5, learning_method='online', learning_offset=50,
                                             random_state=0))
        ]))
    ])),
    ('KBest', SelectKBest(f_classif, k=2500)),
    #('to_dense', DenseTransformer()),    
    #('Percentile', SelectPercentile(f_classif, percentile=0.75))
    #('PCA', PCA(n_components=0.90, svd_solver='full'))
])

In [337]:
tweet_malo = X_train[382]

In [486]:
from scipy.sparse import hstack, vstack, coo_matrix

features_train = textFeatures.fit_transform(X_train, y_train)
features_train.shape

(2234, 3000)

In [487]:
features_test = textFeatures.transform(X_test)
features_test.shape

(222, 3000)

In [522]:
X_ag = df_HS['text'].values
y_ag = df_HS['AG'].values
X_tr = df_HS['text'].values
y_tr = df_HS['TR'].values

In [524]:
print(sorted(Counter(y_ag).items()))

[(0, 399), (1, 1661)]


In [525]:
items = Counter(y_ag)
diff_ag = items[1] - items[0]
diff_ag

1262

In [526]:
items = Counter(y_tr)
diff_tr = items[1] - items[0]
diff_tr

448

In [527]:
# AG
df_dev_extra = df_dev[df_dev['HS'] == 0]
X_extra = df_dev_extra['text'].values
y_extra = df_dev_extra['AG'].values
X_extra = np.random.choice(X_extra, size=diff_ag)
y_extra = y_extra[:diff_ag]
X_ag = np.append(X_ag, X_extra)
y_ag = np.append(y_ag, y_extra)
assert len(X_ag) == len(y_ag)
permutation = np.random.permutation(len(X_ag))
X_ag = X_ag[permutation]
y_ag = y_ag[permutation]

# TR
X_extra = df_dev_extra['text'].values
y_extra = df_dev_extra['TR'].values
X_extra = np.random.choice(X_extra, size=diff_tr)
y_extra = y_extra[:diff_tr]
X_tr = np.append(X_tr, X_extra)
y_tr = np.append(y_tr, y_extra)
assert len(X_tr) == len(y_tr)
permutation = np.random.permutation(len(X_tr))
X_tr = X_tr[permutation]
y_tr = y_tr[permutation]

In [627]:
print(sorted(Counter(y_tr).items()))

[(0, 1254), (1, 1254)]


In [ ]:
features_dev = textFeatures.fit_transform(X_ag, y_ag)
features_dev.shape

In [445]:
% pdb

Automatic pdb calling has been turned OFF


We define our model

### Evaluation

In [115]:
import multiprocessing
n_jobs = int(multiprocessing.cpu_count() * 0.8)

In [488]:
import warnings 
warnings.filterwarnings("ignore")
from evaluation import evaluate
results = evaluate(features_train, y_train, features_test, y_test, classifiers, n_jobs=n_jobs)
results

  0%|          | 0/3 [00:00<?, ?it/s]

logistic_regression
Best score:  0.925252525253
Best params:  {'clf__C': 0.10000000000000001, 'clf__penalty': 'l1', 'clf__tol': 0.0001}




 33%|███▎      | 1/3 [01:13<02:27, 73.94s/it]

linear_SVC
Best score:  0.923744096179
Best params:  {'clf__C': 0.69999999999999996, 'clf__penalty': 'l1', 'clf__tol': 0.1}




 67%|██████▋   | 2/3 [04:12<01:45, 105.29s/it]

random_forest
Best score:  0.925252525253
Best params:  {'clf__max_depth': 8, 'clf__n_estimators': 40}




100%|██████████| 3/3 [05:08<00:00, 90.44s/it] 


{'logistic_regression': {'params': {'memory': None,
   'steps': [('clf',
     LogisticRegression(C=0.10000000000000001, class_weight=None, dual=False,
               fit_intercept=True, intercept_scaling=1, max_iter=100,
               multi_class='warn', n_jobs=None, penalty='l1', random_state=None,
               solver='warn', tol=0.0001, verbose=0, warm_start=False))],
   'clf': LogisticRegression(C=0.10000000000000001, class_weight=None, dual=False,
             fit_intercept=True, intercept_scaling=1, max_iter=100,
             multi_class='warn', n_jobs=None, penalty='l1', random_state=None,
             solver='warn', tol=0.0001, verbose=0, warm_start=False),
   'clf__C': 0.10000000000000001,
   'clf__class_weight': None,
   'clf__dual': False,
   'clf__fit_intercept': True,
   'clf__intercept_scaling': 1,
   'clf__max_iter': 100,
   'clf__multi_class': 'warn',
   'clf__n_jobs': None,
   'clf__penalty': 'l1',
   'clf__random_state': None,
   'clf__solver': 'warn',
   'clf__tol'

### Outputting the results

In [ ]:
from evaluation import createModel
classifier = svm.LinearSVC(dual=False, penalty='l2', tol=0.0001, C=0.8)
#classifier = linear_model.LogisticRegression(penalty='l1', C=0.1, tol=0.0001)

model = createModel(classifier)

In [ ]:
model.fit(features_dev, y_ag)

In [539]:
df_predictions = load_data('es_a.tsv')
df_predictions

id  HS
0     30344   0
1     30466   0
2     31084   0
3     34552   1
4     32538   1
5     32951   0
6     33564   1
7     31367   1
8     34047   0
9     33491   1
10    33064   0
11    32025   1
12    31184   1
13    30102   1
14    31454   1
15    32070   0
16    31241   0
17    30333   1
18    31923   0
19    31966   1
20    30277   0
21    31477   0
22    33750   0
23    31311   1
24    32895   0
25    32600   0
26    33630   1
27    31698   0
28    30561   0
29    33567   1
...     ...  ..
1570  31525   1
1571  32583   0
1572  31544   1
1573  30350   0
1574  33888   1
1575  34338   0
1576  32894   0
1577  32000   1
1578  31902   1
1579  31384   1
1580  34524   0
1581  30355   1
1582  32320   0
1583  32306   0
1584  34199   1
1585  31962   1
1586  33110   1
1587  30459   1
1588  31663   0
1589  34301   0
1590  30987   1
1591  31379   0
1592  33584   1
1593  33535   1
1594  34383   0
1595  32135   1
1596  32116   1
1597  31766   1
1598  34536   1
1599  33712   1

[1600 rows x 2 columns]

In [646]:
df_results = df_evaluation.copy()
df_results.insert(2, 'HS', df_predictions['HS'].values)
df_results['TR'] = pd.Series(np.zeros(df_results.shape[0])-1, dtype='int64')
df_results['AG'] = pd.Series(np.zeros(df_results.shape[0])-1, dtype='int64')

In [ ]:
from tqdm import tqdm
for row in tqdm(df_results.itertuples()):
        df_results.at[row.Index, 'AG'] = 0 if row[3] == 0 else model.predict(textFeatures.transform([row[2]]))

In [647]:
df_results

id                                               text  HS  TR  AG
0     30344  #CadaMañana cállate la puta que me pario Kohan...   0  -1  -1
1     30466                   Estas navidades mi polla mereces   0  -1  -1
2     31084  Si no aprobas te pego una cojida que no te la ...   0  -1  -1
3     34552  @AlecitoGamer @falklands_utd A, no entendiste ...   1  -1  -1
4     32538  @deborahhq1973 Es normal, en Cataluña los Inde...   1  -1  -1
5     32951  @Zurine3 Incluyendo mundo educativo. Conociend...   0  -1  -1
6     33564  Tu eres una hija de puta en la cama que hace p...   1  -1  -1
7     31367  @Kaikot Víctimas son los subsaharianos que vie...   1  -1  -1
8     34047  @itsbreeen Soy la puta ama y lo demás son tont...   0  -1  -1
9     33491  La zorra asiática termina con una abundante co...   1  -1  -1
10    33064  La inmigración ilegal que entra a diario por m...   0  -1  -1
11    32025  Vamos a seguir permitiendo esto? No los querem...   1  -1  -1
12    31184                         Ay wey cállate pinche puta   1  -1  -1
13    30102  @Bohaliciousss Pues es que tu eres la más puta...   1  -1  -1
14    31454  @el_gungui @vox_es @Santi_ABASCAL @voz_populi ...   1  -1  -1
15    32070  El hermano de Arturo Pérez Reverte, ex policía...   0  -1  -1
16    31241  A vos choriplanero y fanático de la política, ...   0  -1  -1
17    30333                   @infobae Cállate....hija de puta   1  -1  -1
18    31923  Que tu haces lo que quieras de tu puta vida. P...   0  -1  -1
19    31966  Agradezco por las pequeñas cosas, por enseñarm...   1  -1  -1
20    30277  @Barbijaputa ....toma que te veo despistada. D...   0  -1  -1
21    31477  LA PUTA MADRE QUE LO PARIO MEL SE MERECÍA TODO...   0  -1  -1
22    33750  New post: "Trump aprovecha el asesinato de una...   0  -1  -1
23    31311  ¿Me podéis dar mas asco los que pensáis que po...   1  -1  -1
24    32895  la primera novia es una puta siempre pero te e...   0  -1  -1
25    32600  @Albert_Rivera @guardiacivil ¿Deseas también p...   0  -1  -1
26    33630                          @ash_london Cállate zorra   1  -1  -1
27    31698  FLOR SALTANDO AHÍ LA RE PUTA MADRE TE LO MEREC...   0  -1  -1
28    30561        SE LO MERECIA TANTO FLORENCIA LA PUTA MADRE   0  -1  -1
29    33567                     chicharra puta cállate y dormí   1  -1  -1
...     ...                                                ...  ..  ..  ..
1570  31525  Mereces morir tu hija de la grandísima puta ht...   1  -1  -1
1571  32583  @nohaydeque Todavía me acuerdo cuando tenía 10...   0  -1  -1
1572  31544  @alvaroaletorres  ya comiste?? Jajajaja pinche...   1  -1  -1
1573  30350  Calláte Rossi, la recalcada concha de la lora ...   0  -1  -1
1574  33888                 @biankitooo Si decore cállate puta   1  -1  -1
1575  34338  Tú apoyo?? Quitando concertinas y recursos a l...   0  -1  -1
1576  32894  CARL GRIMES TE MERECÍAS TODO LA PUTA MSDRE #TW...   0  -1  -1
1577  32000  @Cooperativa Y recien se vienen a dar cuenta? ...   1  -1  -1
1578  31902  Activista alemana proinmigracin cumple su sueo...   1  -1  -1
1579  31384            @lapuntitanamas Negrata con tres patas!   1  -1  -1
1580  34524  Los socialistas han expulsado en caliente a lo...   0  -1  -1
1581  30355  Los últimos saltos de valla por parte de inmig...   1  -1  -1
1582  32320  @Barbijaputa El Sr,lo lees de izquierda a dere...   0  -1  -1
1583  32306     @Barbijaputa A todo, como Madre lo sé!!!! 😂😂😂😂   0  -1  -1
1584  34199  @ElQuemaMarias te parece que un sudaca tatuado...   1  -1  -1
1585  31962  @isaac_c_h @elyyvaliente Subnormal dime que se...   1  -1  -1
1586  33110  Cállate, maldita perra. Nadie cree en ti ni en...   1  -1  -1
1587  30459  Acabo de descubrir que "sudaca" es una palabra...   1  -1  -1
1588  31663  Que caiga todo el peso de la ley y después nad...   0  -1  -1
1589  34301  los invito a la fiesta de nuestros amigos. est...   0  -1  -1
1590  30987  Pa que en la cita la zorra no se alborote, rom...   1  -1  -1
1591  31379  El juez rechaza una d

In [645]:
df_results.drop(columns=['text'], inplace=True)
df_results.to_csv('es_b.tsv', index=False, sep="\t", header=None)